adapters/coap-vertx-base/src/main/java/org/eclipse/hono/adapter/coap/AbstractVertxBasedCoapAdapter.java

```diff
@@ -0,0 +1,419 @@
+ */
+public abstract class AbstractVertxBasedCoapAdapter<T extends CoapAdapterProperties>
+        extends AbstractProtocolAdapterBase<T> {
+
+    protected final Logger log = LoggerFactory.getLogger(getClass());
+
+    /**
+     * COAP server. Created from blocking execution, therefore use volatile.
+     */
+    private CoapServer server;
+    private volatile Endpoint insecureEndpoint;
+    private CoapAdapterMetrics metrics;
+
+    /**
+     * Sets the metrics for this service.
+     *
+     * @param metrics The metrics
+     */
+    @Autowired
+    public final void setMetrics(final CoapAdapterMetrics metrics) {
+        this.metrics = metrics;
+    }
+
+    /**
+     * @return {@link CoAP#DEFAULT_COAP_SECURE_PORT}
+     */
+    @Override
+    public final int getPortDefaultValue() {
+        return CoAP.DEFAULT_COAP_SECURE_PORT;
+    }
+
+    /**
+     * @return {@link CoAP#DEFAULT_COAP_PORT}
+     */
+    @Override
+    public final int getInsecurePortDefaultValue() {
+        return CoAP.DEFAULT_COAP_PORT;
+    }
+
+    @Override
+    protected final int getActualPort() {
+        return Constants.PORT_UNCONFIGURED;
+    }
+
+    @Override
+    protected final int getActualInsecurePort() {
+        int port = Constants.PORT_UNCONFIGURED;
+        final Endpoint endpoint = insecureEndpoint;
+        if (endpoint != null) {
+            port = endpoint.getAddress().getPort();
+        }
+        return port;
+    }
+
+    /**
+     * Called during {@link #doStart(Future)}.
+     * 
+     * Note: The call is in the scope of "executeBlocking" and therefore special care accessing this instance must be
+     * obeyed.
+     * 
+     * @param adapterContext context of this adapter. Intended for schedule calls back into this vertical.
+     * @param server coap server to add resources.
+     */
+    protected abstract void addResources(Context adapterContext, CoapServer server);
+
+    /**
+     * Sets the coap server instance configured to serve requests.
+     * <p>
+     * If no server is set using this method, then a server instance is created during startup of this adapter based on
+     * the <em>config</em> properties.
+     * 
+     * @param server The coap server.
+     * @throws NullPointerException if server is {@code null}.
+     */
+    @Autowired(required = false)
+    public final void setCoapServer(final CoapServer server) {
+        Objects.requireNonNull(server);
+        this.server = server;
+    }
+
+    @Override
+    public final void doStart(final Future<Void> startFuture) {
+        checkPortConfiguration().compose(s -> preStartup()).compose(s -> {
+            final Future<Void> deployFuture = Future.future();
+            // access environment using the context of this vertx and
+            // load it into finals to access them within the executeBlocking
+            final CoapAdapterProperties config = getConfig();
+            final Context adapterContext = getVertx().getOrCreateContext();
```

Comment:
you can simply refer to this adapter's vert.x context by the `context` field, which is set by vert.x on startup of the Verticle.

Reply:
Will that `context` field be visible for other threads?
I could see that in the implementation nor did I find it in the docu.

```
final Context adapterContext = context; 
```

so that would be the resulting solution. Or do you have more information about using the context from other threads?